### PR TITLE
Hides user and credential field from search response

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
- Hides user field from Search and Get APIs
- A painless script was added in https://github.com/opensearch-project/flow-framework/pull/504 to filter out the credential field when present and adds a `filter` field instead of `_source` causing more client side work. Handled the credential field filtering in a better way.

Testing

Request
```json
curl -XPOST --insecure -u 'admin:admin' --header 'Content-Type: application/json' "https://localhost:9200/_plugins/_flow_framework/workflow?provision=true" -d '{"name":"Deploy Claude Model","description":"Deploy a model using a connector to Claude","use_case":"PROVISION","version":{"template":"1.0.0","compatibility":["2.12.0","3.0.0"]},"workflows":{"provision":{"nodes":[{"id":"create_claude_connector","type":"create_connector","user_inputs":{"name":"Claude Instant Runtime Connector","version":"1","protocol":"aws_sigv4","description":"The connector to BedRock service for Claude model","actions":[{"headers":{"x-amz-content-sha256":"required","content-type":"application/json"},"method":"POST","request_body":"{ \"prompt\":\"${parameters.prompt}\", \"max_tokens_to_sample\":${parameters.max_tokens_to_sample}, \"temperature\":${parameters.temperature},  \"anthropic_version\":\"${parameters.anthropic_version}\" }","action_type":"predict","url":"https://bedrock-runtime.us-west-2.amazonaws.com/model/anthropic.claude-instant-v1/invoke"}],"credential":{"access_key":"PUT_YOUR_ACCESS_KEY_HERE","secret_key":"PUT_YOUR_SECRET_KEY_HERE"},"parameters":{"endpoint":"bedrock-runtime.us-west-2.amazonaws.com","content_type":"application/json","auth":"Sig_V4","max_tokens_to_sample":"8000","service_name":"bedrock","temperature":"0.0001","response_filter":"$.completion","region":"us-west-2","anthropic_version":"bedrock-2023-05-31"}}}]}}}'
```

Response of Search API:
```
curl -XGET --insecure -u 'admin:admin' --header 'Content-Type: application/json' "https://localhost:9200/_plugins/_flow_framework/workflow/_search?pretty" -d '{"query":{"match_all":{}}}'
{
  "took" : 44,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : ".plugins-flow-framework-templates",
        "_id" : "YinyEo8B0L4I4ADcw5V9",
        "_version" : 2,
        "_seq_no" : 1,
        "_primary_term" : 1,
        "_score" : 1.0,
        "_source" : {
          "use_case" : "PROVISION",
          "created_time" : 1714009850537,
          "last_updated_time" : 1714009850537,
          "name" : "Deploy Claude Model",
          "description" : "Deploy a model using a connector to Claude",
          "workflows" : {
            "provision" : {
              "nodes" : [
                {
                  "user_inputs" : {
                    "protocol" : "aws_sigv4",
                    "name" : "Claude Instant Runtime Connector",
                    "description" : "The connector to BedRock service for Claude model",
                    "version" : "1",
                    "parameters" : {
                      "endpoint" : "bedrock-runtime.us-west-2.amazonaws.com",
                      "content_type" : "application/json",
                      "auth" : "Sig_V4",
                      "max_tokens_to_sample" : "8000",
                      "service_name" : "bedrock",
                      "temperature" : "0.0001",
                      "response_filter" : "$.completion",
                      "region" : "us-west-2",
                      "anthropic_version" : "bedrock-2023-05-31"
                    },
                    "actions" : [
                      {
                        "headers" : {
                          "x-amz-content-sha256" : "required",
                          "content-type" : "application/json"
                        },
                        "method" : "POST",
                        "request_body" : "{ \"prompt\":\"${parameters.prompt}\", \"max_tokens_to_sample\":${parameters.max_tokens_to_sample}, \"temperature\":${parameters.temperature},  \"anthropic_version\":\"${parameters.anthropic_version}\" }",
                        "action_type" : "predict",
                        "url" : "https://bedrock-runtime.us-west-2.amazonaws.com/model/anthropic.claude-instant-v1/invoke"
                      }
                    ]
                  },
                  "id" : "create_claude_connector",
                  "type" : "create_connector",
                  "previous_node_inputs" : { }
                }
              ],
              "edges" : [ ],
              "user_params" : { }
            }
          },
          "version" : {
            "template" : "1.0.0",
            "compatibility" : [
              "2.12.0",
              "3.0.0"
            ]
          },
          "last_provisioned_time" : 1714009850887
        }
      }
    ]
  }
}
```

Response of GET API:

```
curl -XGET --insecure -u 'admin:admin' --header 'Content-Type: application/json' "https://localhost:9200/_plugins/_flow_framework/workflow/D0sjE48BLSYJV6YQCQ_K?pretty" 
{
  "name" : "test-1",
  "description" : "test-2",
  "use_case" : "SEMANTIC_SEARCH",
  "version" : {
    "template" : "1.0.0",
    "compatibility" : [
      "2.12.0",
      "3.0.0"
    ]
  },
  "workflows" : {
    "provision" : {
      "user_params" : { },
      "nodes" : [
        {
          "id" : "create_ingest_pipeline",
          "type" : "create_ingest_pipeline",
          "previous_node_inputs" : { },
          "user_inputs" : {
            "configurations" : "{\"description\":\"test-pipeline\",\"processors\":[{\"text_embedding\":{\"field_map\":{\"passage_text\":\"passage_embedding\"},\"model_id\":\"jyKh544B161srMZWwY93\"}}]}",
            "pipeline_id" : "nlp-pipeline"
          }
        }
      ],
      "edges" : [ ]
    }
  },
  "user" : {
    "name" : "admin",
    "backend_roles" : [
      "admin"
    ],
    "roles" : [
      "own_index",
      "all_access"
    ],
    "custom_attribute_names" : [ ],
    "user_requested_tenant" : null
  },
  "created_time" : 1714013014339,
  "last_updated_time" : 1714013014339,
  "last_provisioned_time" : 1714013014615
}
```

### Issues Resolved
Resolves https://github.com/opensearch-project/flow-framework/issues/546

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java [hunk 1]
```diff
@@ -204,9 +204,29 @@ String decrypt(final String encryptedCredential) {
...
+                .useCase(template.useCase())
+                .templateVersion(template.templateVersion())
+                .user(null)
+                .uiMetadata(template.getUiMetadata())
+                .compatibilityVersion(template.compatibilityVersion())
+                .workflows(template.workflows())
+                .createdTime(template.createdTime())
+                .lastUpdatedTime(template.lastUpdatedTime())
+                .lastProvisionedTime(template.lastProvisionedTime())
+                .build();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java:222]: The Builder constructor using the old template already copies over the other fields
```suggestion
            updatedTemplate = new Template.Builder(template)
                .user(null)
                .build();
```

File: src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java [hunk 2]
```diff
@@ -204,9 +204,29 @@ String decrypt(final String encryptedCredential) {
...
+        } else {
+            updatedTemplate = template;
+        }
+
         Map<String, Workflow> processedWorkflows = new HashMap<>();
-        for (Map.Entry<String, Workflow> entry : template.workflows().entrySet()) {
+
+        for (Map.Entry<String, Workflow> entry : updatedTemplate.workflows().entrySet()) {
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java:229]: These are the same workflows as before.  Seems we could move this workflow redaction step above the user check and replace both fields (user always with null and new workflows).

File: src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java (hunk 1 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java]: Ah! I missed to change this after the testing

File: src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java (hunk 2 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java]: Not sure if I understand this correctly but I moved the user check to later. Let me know if you meant something else and I will make the changes

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java]: What I meant to say was that `template.workflows()` and `updatedTemplate.workflows()` are exactly the same, as the update only (possibly) changed user, so this particular variable change doesn't really mean anything (although it's more technically correct).

There are two steps: (1) conditionally replace user with null if it's non-null, and (2) always replace workflows.  Those steps could be combined into one step replacing user (always null even if it was before) and workflows.

File: src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java [hunk 3]
```diff
@@ -227,7 +228,15 @@ public Template redactTemplateCredentials(Template template) {
...
-        return new Template.Builder(template).workflows(processedWorkflows).build();
+        Template updatedTemplate;
+
+        if (template.getUser() != null) {
+            updatedTemplate = new Template.Builder(template).name(template.name()).user(null).build();
+        } else {
+            updatedTemplate = template;
+        }
+
+        return new Template.Builder(updatedTemplate).workflows(processedWorkflows).build();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java:239]:
```suggestion
        return new Template.Builder(template).user(null),workflows(processedWorkflows).build();
```

Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java:239]: Wouldn't this add a user field in the builder, although null but even when it's not present?

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java:239]: There already is a user field initialized thus:
```java
private User user = null;
```
So no difference here with an uninitialized vs. intentionally cleared.

Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java]: Ah! Sure, that makes sense. I like the one line change now for this method.

**Review by dbwiddis — APPROVED** [review]

**Comment by dbwiddis** [comment]
Approved, but it doesn't look like tests are covering the `getSourceContext` static method in either case. Are these covered in integ tests?  Also it looks like the TODO has been handled.

File: src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java [hunk 4]
```diff
@@ -19,8 +20,16 @@
  */
 public class RestHandlerUtils {
 
+    /** Path to credential field **/
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java:23]: Will this work even if create connector node is last node in template?

File: src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java [hunk 5]
```diff
@@ -79,8 +79,8 @@ protected void doExecute(Task task, WorkflowRequest request, ActionListener<GetW
                         logger.error(errorMessage);
                         listener.onFailure(new FlowFrameworkException(errorMessage, RestStatus.NOT_FOUND));
                     } else {
-                        // Remove any credential from response
-                        Template template = encryptorUtils.redactTemplateCredentials(Template.parse(response.getSourceAsString()));
+                        // Remove any secured field from response
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java:82]: Can we do this only if the user isn't admin so there is still some way to look at user field, similar to AD we have this check for a similar use case:
https://github.com/opensearch-project/anomaly-detection/blob/f3998e9101d1cd1c246b30756ee7d3a3ca06c757/src/main/java/org/opensearch/timeseries/util/ParseUtils.java#L592

File: src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java [hunk 6]
```diff
@@ -35,10 +44,11 @@ public static FetchSourceContext getSourceContext(RestRequest request, SearchSou
         // TODO
         // 1. check if the request came from dashboard and exclude UI_METADATA
         if (searchSourceBuilder.fetchSource() != null) {
-            String[] newArray = (String[]) ArrayUtils.addAll(searchSourceBuilder.fetchSource().excludes(), USER_EXCLUDE);
+            String[] newArray = (String[]) ArrayUtils.addAll(searchSourceBuilder.fetchSource().excludes(), DASHBOARD_EXCLUDES);
             return new FetchSourceContext(true, searchSourceBuilder.fetchSource().includes(), newArray);
         } else {
-            return null;
+            // When user does not set the _source field in search api request, searchSourceBuilder.fetchSource becomes null
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java:50]: shouldn't we want to exclude still and not have an empty array here?

File: src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java (hunk 5 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java:82]: Also we still want user for the state index to distinguish who provisions.

Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java:82]: yes agree but still think only admin should be able to see the user to see who provisioned it, which is what is done in alerting. What do you think about that?

Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java:82]: > yes agree but still think only admin should be able to see the user to see who provisioned it, which is what is done in alerting.

+1

File: src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java (hunk 4 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java:23]: Why wouldn't it? We are adding the filter on the SearchSourceBuilder itself, order doesn't matter. 
To be more assured tested with the below template

```json
{
    "name": "Deploy Claude Model",
    "description": "Deploy a model using a connector to Claude",
    "use_case": "PROVISION",
    "version": {
        "template": "1.0.0",
        "compatibility": ["2.12.0", "3.0.0"]
    },
    "workflows": {
        "provision": {
            "nodes": [{
                "id": "create_ingest_pipeline",
                "type": "create_ingest_pipeline",
                "user_inputs": {
                    "pipeline_id": "nlp-pipeline",
                    "configurations": {
                        "description": "test-pipeline",
                        "processors": [{
                            "text_embedding": {
                                "model_id": "jyKh544B161srMZWwY93",
                                "field_map": {
                                    "passage_text": "passage_embedding"
                                }
                            }
                        }]
                    }
                }
            }, {
                "id": "create_claude_connector",
                "type": "create_connector",
                "user_inputs": {
                    "name": "Claude Instant Runtime Connector",
                    "version": "1",
                    "protocol": "aws_sigv4",
                    "description": "The connector to BedRock service for Claude model",
                    "actions": [{
                        "headers": {
                            "x-amz-content-sha256": "required",
                            "content-type": "application/json"
                        },
                        "method": "POST",
                        "request_body": "{ \"prompt\":\"${parameters.prompt}\", \"max_tokens_to_sample\":${parameters.max_tokens_to_sample}, \"temperature\":${parameters.temperature},  \"anthropic_version\":\"${parameters.anthropic_version}\" }",
                        "action_type": "predict",
                        "url": "https://bedrock-runtime.us-west-2.amazonaws.com/model/anthropic.claude-instant-v1/invoke"
                    }],
                    "credential": {
                        "access_key": "PUT_YOUR_ACCESS_KEY_HERE",
                        "secret_key": "PUT_YOUR_SECRET_KEY_HERE"
                    },
                    "parameters": {
                        "endpoint": "bedrock-runtime.us-west-2.amazonaws.com",
                        "content_type": "application/json",
                        "auth": "Sig_V4",
                        "max_tokens_to_sample": "8000",
                        "service_name": "bedrock",
                        "temperature": "0.0001",
                        "response_filter": "$.completion",
                        "region": "us-west-2",
                        "anthropic_version": "bedrock-2023-05-31"
                    }
                }
            }]
        }
    }
}
```

Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java:23]: meant to type to add test cases more so, I am okay with manual test too so we can get this in

File: src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java (hunk 5 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/transport/GetWorkflowTransportAction.java:84]: Added the user check for Get API, Search API and Status API.

**Comment by owaiskazi19** [comment]
@dbwiddis you might want to review again since I added more changes. Thanks

File: src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java [hunk 7]
```diff
@@ -19,26 +20,38 @@
...
      * @param searchSourceBuilder the search request source builder
      * @return modified sources
      */
-    public static FetchSourceContext getSourceContext(RestRequest request, SearchSourceBuilder searchSourceBuilder) {
+    public static FetchSourceContext getSourceContext(User user, SearchSourceBuilder searchSourceBuilder) {
         // TODO
         // 1. check if the request came from dashboard and exclude UI_METADATA
         if (searchSourceBuilder.fetchSource() != null) {
-            String[] newArray = (String[]) ArrayUtils.addAll(searchSourceBuilder.fetchSource().excludes(), USER_EXCLUDE);
+            String[] newArray = (String[]) ArrayUtils.addAll(searchSourceBuilder.fetchSource().excludes(), DASHBOARD_EXCLUDES);
```
Comment (amitgalitz) [inline src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java:47]: Do we have actual logic to check dashboard_excludes

**Comment by owaiskazi19** [comment]
> Approved, but it doesn't look like tests are covering the `getSourceContext` static method in either case. Are these covered in integ tests? Also it looks like the TODO has been handled.

Added tests for RestHandlerUtils.

File: src/main/java/org/opensearch/flowframework/util/ParseUtils.java [hunk 8]
```diff
@@ -113,6 +110,18 @@ public static void buildStringToStringMap(XContentBuilder xContentBuilder, Map<?
...
     }
 
+    /**
+     * 'all_access' role users are treated as admins.
+     * @param user of the current role
+     * @return boolean if the role is admin
+     */
+    public static boolean isAdmin(User user) {
+        if (user == null) {
+            return false;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/util/ParseUtils.java:120]: If security is disabled all users show up as null, correct?

**Review by dbwiddis — APPROVED** [review]
(Re)approved with a question.

File: src/main/java/org/opensearch/flowframework/util/ParseUtils.java (hunk 8 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/util/ParseUtils.java:120]: Yes

File: src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java (hunk 7 above)
Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/util/RestHandlerUtils.java:45]: talked offline that if we have fetchSource there in the request its a sign the frontend made the request.

**Review by amitgalitz — APPROVED** [review]

**Pull request merged**